modules/core/src/main/scala/org/scalasteward/core/data/Version.scala

```diff
@@ -157,61 +149,34 @@ object Version {
           case _                               => 0
         }
     }
+    final case class Hash(value: String, startIndex: Int) extends Component
     final case class Separator(c: Char, startIndex: Int) extends Component
     case object Empty extends Component { override def startIndex: Int = -1 }
 
-    def parse(str: String): List[Component] = {
-      @tailrec
-      def loop(
-          rest: List[Char],
-          accN: List[Char],
-          accA: List[Char],
-          acc: List[Component],
-          index: Int
-      ): List[Component] =
-        rest match {
-          case h :: t =>
-            h match {
-              case '.' | '-' | '_' | '+' =>
-                val (newAcc, newIndex) = materialize(accN, accA, acc, index)
-                loop(t, List.empty, List.empty, Separator(h, newIndex) :: newAcc, newIndex + 1)
-
-              case _ if h.isDigit && accA.nonEmpty =>
-                val (newAcc, newIndex) = materialize(List.empty, accA, acc, index)
-                loop(t, h :: accN, List.empty, newAcc, newIndex)
-
-              case _ if h.isDigit && accA.isEmpty =>
-                loop(t, h :: accN, List.empty, acc, index)
-
-              case _ if accN.nonEmpty =>
-                val (newAcc, newIndex) = materialize(accN, List.empty, acc, index)
-                loop(t, List.empty, h :: accA, newAcc, newIndex)
-
-              case _ if accN.isEmpty =>
-                loop(t, List.empty, h :: accA, acc, index)
-            }
-          case Nil =>
-            val (newAcc, _) = materialize(accN, accA, acc, index)
-            newAcc
-        }
+    private val numeric = """^([0-9]+).*""".r
+    private val separator = """^([.\-_+]).*""".r
+    private val alpha = """^([^.\-_+0-9]+).*""".r
+    private val hash = """([-+])(g?\p{XDigit}{6,}).*""".r
+
+    def parse(str: String, index: Int = 0): List[Component] =
+      str match {
+        case "" => List.empty
+        case hash(sep, value) if !startsWithDate(value) =>
+          Separator(sep.head, index) +: Hash(value, index + 1) +:
+            parse(str.substring(value.length + 1), index + value.length + 1)
+        case numeric(value)   => Numeric(value, index) +: rest(value, str, index)
+        case alpha(value)     => Alpha(value, index) +: rest(value, str, index)
+        case separator(value) => Separator(value.head, index) +: rest(value, str, index)
+      }
```

Comment:
This looks great @mzuehlke! I wonder if it could further simplified by replacing the `rest` function with capturing the rest in each regex, i.e. `^([0-9]+)(.*)` instead of `^([0-9]+).*`.  WDYT?